### PR TITLE
feat(tests) add EL-withdrawal unit tests (fix issue #536, BS-4018)

### LINKED ISSUE #536: Missing tests for river
```
  Significant gaps:

  ┌─────────────────────────────────────────────┬──────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
  │                     Gap                     │                                                      Detail                                                      │
  ├─────────────────────────────────────────────┼──────────────────────────────────────────────────────────────────────────────────────────────────────────────────┤
  │ All fee/value logic                         │ Every partial test calls requestETHExits(empty, allocs, 0) — maxFee = 0 and no msg.value. So fee forwarding, the │
  │                                             │  excess-refund branch (:286-292), and UnsentRefund have zero coverage. This is the part that moves real ETH.     │
  ├─────────────────────────────────────────────┼──────────────────────────────────────────────────────────────────────────────────────────────────────────────────┤
  │                                             │ MockWithdrawForPartialExits skips all of WithdrawV1's validations. testPartialExitEmptyAmountsIsNoOp passes      │
  │ Mock diverges from real Withdraw            │ empty pubkeys/amounts and asserts a no-op — but the real Withdraw.withdraw reverts InvalidEmptyArray on empty    │
  │                                             │ pubkeys. That test encodes production-invalid behavior. No test wires the real WithdrawV1 into the registry.     │
  ├─────────────────────────────────────────────┼──────────────────────────────────────────────────────────────────────────────────────────────────────────────────┤
  │                                             │ Not referenced by any test (tests use sudoSetWithdrawAddress). The actual migration path that sets               │
  │ initOperatorsRegistryV1_2(_withdrawAddress) │ WithdrawAddress is uncovered — verify the V2→V3 migration fork tests were updated to the new signature (and      │
  │                                             │ still compile).                                                                                                  │
  ├─────────────────────────────────────────────┼──────────────────────────────────────────────────────────────────────────────────────────────────────────────────┤
  │ Mixed full + partial in one call            │ No test passes both arrays non-empty — the combined requestedETHAmount vs demand path and same-operator-in-both  │
  │                                             │ reservation are untested.                                                                                        │
  ├─────────────────────────────────────────────┼──────────────────────────────────────────────────────────────────────────────────────────────────────────────────┤
  │ Partial AllocationWithIncorrectAmount       │ The per-amount gweiAmount < 1 gwei revert (:352) is untested (only the full-exit < MIN_ETH_AMOUNT is).           │
  ├─────────────────────────────────────────────┼──────────────────────────────────────────────────────────────────────────────────────────────────────────────────┤
  │ Event assertion                             │ RequestedPartialETHExits is never asserted via expectEmit.                                                       │
  ├─────────────────────────────────────────────┼──────────────────────────────────────────────────────────────────────────────────────────────────────────────────┤
  │ Length mismatch                             │ pubkeys.length != amounts.length (reverts in real Withdraw) untested.                                            │
```

_Originally posted by @mischat in https://github.com/liquid-collective/liquid-collective-protocol/pull/372#discussion_r3337571898_

### PR DESCRIPTION
**Pull request by juliaaschmidt**
closes issue #536 

## Description

This pull request enhances the EL exit tests for the `OperatorsRegistry` by introducing more realistic scenarios involving the real withdrawal contract, improving coverage for fee handling, refunds, and event emissions. It also adds utility contracts and helper functions to streamline test setup and simulate edge cases.

**Withdraw integration and fee/refund handling:**
- Added integration tests using the real `WithdrawV1` contract, including a helper (`_useRealWithdraw`) to set it up with a mock EL withdrawal contract and fee configuration. Tests now verify correct fee forwarding, refunding of excess ETH, and proper handling when refunds fail (using a new `RejectingRefundRecipient` contract). [[1]](diffhunk://#diff-d676752bfc62e0f8383e9785a7a6b7111cb8e0658b4e32648929d626f6399a21R12-R15) [[2]](diffhunk://#diff-d676752bfc62e0f8383e9785a7a6b7111cb8e0658b4e32648929d626f6399a21R1904-R1918) [[3]](diffhunk://#diff-d676752bfc62e0f8383e9785a7a6b7111cb8e0658b4e32648929d626f6399a21R1944-R1964) [[4]](diffhunk://#diff-d676752bfc62e0f8383e9785a7a6b7111cb8e0658b4e32648929d626f6399a21R2337-R2412)

**Test coverage improvements:**
- Replaced the test for empty amounts being a no-op with a new test ensuring the real withdrawal contract correctly reverts when called with empty pubkey arrays, matching production behavior. [[1]](diffhunk://#diff-d676752bfc62e0f8383e9785a7a6b7111cb8e0658b4e32648929d626f6399a21L2053-R2089) [[2]](diffhunk://#diff-d676752bfc62e0f8383e9785a7a6b7111cb8e0658b4e32648929d626f6399a21R2099-L2073)

**Event and mixed-exit testing:**
- Added tests to verify the emission of the `RequestedELETHExits` event and to confirm that the registry supports handling both CL and EL exits for the same operator in a single request. [[1]](diffhunk://#diff-d676752bfc62e0f8383e9785a7a6b7111cb8e0658b4e32648929d626f6399a21R1929-R1930) [[2]](diffhunk://#diff-d676752bfc62e0f8383e9785a7a6b7111cb8e0658b4e32648929d626f6399a21R2337-R2412)

**Test infrastructure and utilities:**
- Introduced helper functions (`_setupSingleOperator`, `_makeELAlloc`) to simplify and standardize test setup for operators and allocations.
- Added the `RejectingRefundRecipient` contract to simulate refund failures.

These changes provide more robust and production-realistic testing for EL exit flows, fee management, and error handling in the `OperatorsRegistry`.

## Notice

- [ ] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/liquid-collective/liquid-collective-protocol/pulls) for the same update/change?
- [ ] Have you assigned this PR to yourself?
- [ ] Have you added at least 1 reviewer?
- [ ] Have you updated the official documentation?
- [ ] Have you added sufficient documentation in your code?
- [ ] Have you added relevant tests to the official test suite?

## Pull Request Type

- [ ] 💫 New Feature (Breaking Change)
- [ ] 💫 New Feature (Non-breaking Change)
- [ ] 🛠️ Bug fix (Non-breaking Change: Fixes an issue)
- [ ] 🕹️ Chore (Non-breaking Change: Doc updates, pkg upgrades, typos, etc..)

## Testing

- [ ] Have you tested this code with the official test suite?
- [ ] Have you tested this code manually?

## Summary by CodeRabbit

* **Tests**
  * Enhanced test coverage for EL exit (partial-exit) functionality, including scenarios for fee forwarding, refund handling, and mixed CL + EL exits for the same operator.
  * Added validation tests for withdrawal behavior and event emission during exit requests.
  * Improved error handling test coverage for refund failures.

### CODE REVIEW

**Comment by mischat** [comment]
@coderabbitai review

File: contracts/test/OperatorsRegistry.1.t.sol [hunk 1]
```diff
@@ -2307,4 +2334,80 @@ contract OperatorsRegistryV1ELExitTests is Test {
...
+    }
+
+    function testELExitEmitsRequestedELETHExits() public {
+        _setupSingleOperator(32 ether, 32 ether, 8 ether);
+
+        IOperatorsRegistryV1.ExitETHAllocation[] memory empty = new IOperatorsRegistryV1.ExitETHAllocation[](0);
+        IOperatorsRegistryV1.ELExitETHAllocation[] memory allocs = _makeELAlloc(0, EIGHT_ETH_IN_GWEI);
+
+        vm.expectEmit(true, false, false, true, address(reg));
+        emit RequestedELETHExits(0, allocs[0].pubkeys, allocs[0].amounts);
```
[the remark this answers is not in the document]
Comment (juliaaschmidt) [reply contracts/test/OperatorsRegistry.1.t.sol]: Addressed the interface-qualified part — the test now emits `IOperatorsRegistryV1.RequestedELETHExits` and the redundant local event declaration was removed.

On the topic check: `RequestedELETHExits(uint256 indexed index, ...)` has a single indexed param, so `index` is **topic1**, not topic2. `vm.expectEmit(true, false, false, true, address(reg))` already verifies it — the first boolean is `checkTopic1`. The second boolean controls topic2, which this event does not have, so leaving it `false` is correct and consistent with the `FundedValidatorKeys` assertions elsewhere in this file.

File: contracts/test/OperatorsRegistry.1.t.sol [hunk 2]
```diff
@@ -2050,14 +2083,10 @@ contract OperatorsRegistryV1ELExitTests is Test {
...
-    /// Empty amounts array per allocation is a no-op: demand unchanged, requestedExits stays 0.
-    function testELExitEmptyAmountsIsNoOp() public {
-        vm.prank(admin);
-        reg.addOperator("Op0", makeAddr("op0addr"));
-        reg.sudoSetFundedV3(0, 32 ether);
-        reg.sudoSetActiveCLETH(0, 32 ether);
-        vm.prank(river);
-        reg.demandETHExits(8 ether, 32 ether);
+    /// Empty amounts are production-invalid: the real withdrawal contract rejects empty pubkeys.
+    function testELExitWithRealWithdrawRevertsOnEmptyAmounts() public {
```
[the remark this answers is not in the document]
Comment (juliaaschmidt) [reply contracts/test/OperatorsRegistry.1.t.sol]: Fixed — the comment now reads: "An empty EL allocation is production-invalid: the withdrawal contract reverts with `InvalidEmptyArray` when forwarded an empty pubkeys array."

File: contracts/test/OperatorsRegistry.1.t.sol [hunk 3]
```diff
@@ -2413,4 +2440,80 @@ contract OperatorsRegistryV1ELExitTests is Test {
...
+        assertEq(pectraWithdrawal.balance, actualFee, "only the actual withdrawal fee should be paid");
+        assertEq(address(realWithdraw).balance, 0, "withdraw should not retain the max-fee excess");
+        assertEq(address(reg).balance, 0, "registry should not retain caller excess");
+        assertEq(
+            keeper.balance,
+            keeperBalanceBefore - actualFee,
+            "keeper should only spend the actual fee after both refunds"
+        );
+        assertEq(reg.getOperator(0).requestedExits, 8 ether, "requestedExits should reserve the EL amount");
+    }
```
[the remark this answers is not in the document]
Comment (juliaaschmidt) [reply contracts/test/OperatorsRegistry.1.t.sol:2657]: Resolved. The "line 2067" dependency was the `initWithdrawV1_1` call using the old 3-arg signature. It now passes the 4th `_attestationVerifier` argument matching the v2/pectra signature, so this test compiles and passes.

File: contracts/test/OperatorsRegistry.1.t.sol [hunk 4]
```diff
@@ -2413,4 +2440,80 @@ contract OperatorsRegistryV1ELExitTests is Test {
...
+        _setupSingleOperator(32 ether, 32 ether, 8 ether);
+
+        IOperatorsRegistryV1.ExitETHAllocation[] memory empty = new IOperatorsRegistryV1.ExitETHAllocation[](0);
+        IOperatorsRegistryV1.ELExitETHAllocation[] memory allocs = _makeELAlloc(0, EIGHT_ETH_IN_GWEI);
+
+        vm.deal(keeper, valueSent);
+        vm.prank(keeper);
+        vm.expectRevert(abi.encodeWithSelector(IOperatorsRegistryV1.UnsentRefund.selector, keeper, valueSent - maxFee));
+        reg.requestETHExits{value: valueSent}(empty, allocs, maxFee);
+    }
```
[the remark this answers is not in the document]
Comment (juliaaschmidt) [reply contracts/test/OperatorsRegistry.1.t.sol:2675]: Resolved via the same 4-arg `initWithdrawV1_1` fix. Note the suggested source change already exists in production: `requestETHExits` performs the refund with a low-level `call`, checks the boolean success, and reverts `UnsentRefund(msg.sender, excess)` (contracts/src/OperatorsRegistry.1.sol:335-339) — exactly what this test asserts. No source change is needed; this is a test-only PR.

**Review by mischat — APPROVED** [review]
LGTM
